File: results/0.75/ARP/arp-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *create_arp_parser() {
    // Define ARP parsing components
    HParser *h_type = h_uint16();
    HParser *p_type = h_uint16();
    HParser *hlen = h_uint8();
    HParser *plen = h_uint8();
    HParser *oper = h_uint16();
    
    // Capture the length values
    HParser *sha_len = h_bind(hlen, h_repeat_n, h_uint8());
    HParser *spa_len = h_bind(plen, h_repeat_n, h_uint8());
    HParser *tha_len = h_bind(hlen, h_repeat_n, h_uint8());
    HParser *tpa_len = h_bind(plen, h_repeat_n, h_uint8());

    // Construct ARP packet parser
    HParser *arp_parser = h_sequence(
        h_type, // Hardware Type
        p_type, // Protocol Type
        hlen,   // Hardware Address Length
        plen,   // Protocol Address Length
        oper,   // Operation
        sha_len, // Sender Hardware Address
        spa_len, // Sender Protocol Address
        tha_len, // Target Hardware Address
        tpa_len, // Target Protocol Address
        NULL
    );

    return arp_parser;
}

void parse_arp(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *data = (unsigned char *)malloc(file_size);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(data, 1, file_size, file);
    fclose(file);

    HParser *arp_parser = create_arp_parser();
    HParseResult *result = h_parse(arp_parser, data, file_size);

    if (result) {
        printf("ARP packet parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        fprintf(stderr, "Failed to parse ARP packet.\n");
    }

    h_parser_unref(arp_parser);
    free(data);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_arp(argv[1]);
    return EXIT_SUCCESS;
}

File: results/0.75/ARP/arp-gpt-4o.rs
use std::env;
use std::fs::File;
use std::io::Read;
use nom::{
    IResult,
    number::complete::{be_u16, be_u8},
    bytes::complete::take
};

#[derive(Debug)]
struct ArpPacket {
    htype: u16,
    ptype: u16,
    hlen: u8,
    plen: u8,
    oper: u16,
    sha: Vec<u8>,
    spa: Vec<u8>,
    tha: Vec<u8>,
    tpa: Vec<u8>,
}

fn parse_arp(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, htype) = be_u16(input)?;
    let (input, ptype) = be_u16(input)?;
    let (input, hlen) = be_u8(input)?;
    let (input, plen) = be_u8(input)?;
    let (input, oper) = be_u16(input)?;
    
    let sha_len = hlen as usize;
    let spa_len = plen as usize;
    let (input, sha) = take(sha_len)(input)?;
    let (input, spa) = take(spa_len)(input)?;
    let (input, tha) = take(sha_len)(input)?;
    let (input, tpa) = take(spa_len)(input)?;

    Ok((input, ArpPacket {
        htype,
        ptype,
        hlen,
        plen,
        oper,
        sha: sha.to_vec(),
        spa: spa.to_vec(),
        tha: tha.to_vec(),
        tpa: tpa.to_vec(),
    }))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <input file>", args[0]);
        std::process::exit(1);
    }

    let filename = &args[1];
    let mut file = File::open(filename).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_arp(&buffer) {
        Ok((_, packet)) => {
            println!("{:?}", packet);
        }
        Err(err) => {
            eprintln!("Failed to parse ARP packet: {:?}", err);
        }
    }
}